domain-mapping/api/DarkMatter_Domains.php

```diff
@@ -0,0 +1,510 @@
+            }
+        }
+
+        $_domain = array(
+            'active'     => ( ! $active ? false : true ),
+            'blog_id'    => get_current_blog_id(),
+            'domain'     => $fqdn,
+            'is_primary' => ( ! $is_primary ? false : true ),
+            'is_https'   => ( ! $is_https ? false : true ),
+        );
+
+        $result = $this->wpdb->insert( $this->dm_table, $_domain, array(
+            '%d', '%d', '%s', '%d', '%d',
+        ) );
+
+        if ( $result ) {
+            /**
+             * Create the cache key.
+             */
+            $cache_key = md5( $fqdn );
+
+            /**
+             * Update the domain object prior to priming the cache for both the
+             * domain object and the primary domain if necessary.
+             */
+            $_domain['id'] = $this->wpdb->insert_id;
+            wp_cache_add( $cache_key, $_domain, 'dark-matter' );
+
+            if ( $is_primary ) {
+                $dm_primary->set( get_current_blog_id(), $fqdn );
+            }
+
+            $dm_domain = new DM_Domain( (object) $_domain );
+
+            /**
+             * Fire action when a domain is added.
+             *
+             * Fires after a domain is successfully added to the database. This
+             * is also post insertion to the cache.
+             *
+             * @since 2.0.0
+             *
+             * @param DM_Domain $dm_domain Domain object of the newly added Domain.
+             */
+            do_action( 'darkmatter_domain_add', $dm_domain );
+
+            return $dm_domain;
+        }
+
+        return new WP_Error( 'unknown', __( 'Sorry, the domain could not be added. An unknown error occurred.', 'dark-matter' ) );
+    }
+
+    /**
+     * Delete a domain for a specific Site in WordPress.
+     *
+     * @param  string           $fqdn FQDN to be deleted.
+     * @return WP_Error|boolean       True on success. False otherwise.
+     */
+    public function delete( $fqdn = '', $force = true ) {
+        $fqdn = $this->_basic_check( $fqdn );
+
+        if ( is_wp_error( $fqdn ) ) {
+            return $fqdn;
+        }
+
+        /**
+         * Cannot delete what does not exist.
+         */
+        if ( ! $this->is_exist( $fqdn ) ) {
+            return new WP_Error( 'exists', __( 'The domain cannot be found.', 'dark-matter' ) );
+        }
+
+        /**
+         * Check to make sure the domain is assigned to the site.
+         */
+        $_domain = $this->get( $fqdn );
+
+        if ( ! $_domain || get_current_blog_id() !== $_domain->blog_id ) {
+            return new WP_Error( 'not found', __( 'The domain cannot be found.', 'dark-matter' ) );
+        }
+
+        /**
+         * Check to make sure that the domain is not a primary and if it is that
+         * the force flag has been provided.
+         */
+        if ( $_domain->is_primary ) {
+            if ( $force ) {
+                DarkMatter_Primary::instance()->unset();
```

Comment:
Code inspection says that a null pointer exception may occur here and in similar constructions across the plugin (8 more)